### PR TITLE
CMCL-1252: Data sharing between samples and samples can have package dependencies

### PR DESCRIPTION
**Pull request by gaborkb**
### Purpose of this PR
[CMCL-1252](https://jira.unity3d.com/browse/CMCL-1252) Enable data sharing between samples.
Samples can define their asset dependencies. Currently set to the Shared folder. Samples can have sample specific asset dependencies (if ever we need that, e.g. two sample groups have some common assets, these assets don't need to be shared with everyone).

[CMCL-1255](https://jira.unity3d.com/browse/CMCL-1255) Enable samples to have their own package dependencies.
Samples can define their own package dependencies.

Main things to look at:
- com.unity.cinemachine/Editor/Samples/SampleDependencyImporter.cs
- com.unity.cinemachine/Samples~/samples.json
- com.unity.cinemachine/package.json

### Testing status

- [ ] Added an automated test
- [ ] Passed all automated tests
- [x] Manually tested 

### Documentation status

- [ ] Updated [CHANGELOG](https://keepachangelog.com/en/1.0.0/)
- [ ] Updated README (if applicable)
- [ ] Commented all public classes, properties, and methods
- [ ] Updated user documentation

### Comments to reviewers
- Samples define their needs using the `samples.json` found in Samples~ folder.
- Separated our 3D and 2D sample scenes into 2D and 3D samples. This was mainly to test and showcase what this PR does. Could be reverted, and we can come up with better sub-groups (e.g. game samples, film samples, complex samples ...).
- Physics samples is also there to showcase the sample package dependency. Could be removed.

### CODE REVIEW

File: com.unity.cinemachine/Samples~/samples.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+  "SharedAssetDependencies": [
+    "Shared"
```
Comment (glabute) [inline com.unity.cinemachine/Samples~/samples.json:3]: Why have this?  Why not just mention Shared in the package-specific asset dependencies?  Right now, you can't have samples that don't depend on Shared (such as OldSamplesFromCM2)

File: com.unity.cinemachine/Samples~/samples.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      "AssetDependencies": [],
+      "PackageDependencies": []
+    },
+    {
+      "Path": "3D Samples",
+      "AssetDependencies": [],
+      "PackageDependencies": []
+    },
+    {
+      "Path": "Physics Samples",
```
Comment (glabute) [inline com.unity.cinemachine/Samples~/samples.json:17]: I think this is not a good idea.  Our shared scripts have physics dependencies (and there are colliders in the non-physics scenes).  I think it's easier just to accept that our samples all depend on physics.

**Review by glabute — CHANGES_REQUESTED** [review]

**Comment by glabute** [comment]
I like the 2D/3D separation, but I think the Physics separation is going too far

File: com.unity.cinemachine/Samples~/samples.json (hunk 2 above)
Comment (gaborkb) [reply com.unity.cinemachine/Samples~/samples.json:17]: Yes, I agree. I mention that in the Comments to Reviewers part. I removed Physics samples in the next PR - https://github.com/Unity-Technologies/com.unity.cinemachine/pull/665.

**Comment by glabute** [comment]
Now you just need to unseparate the physics samples and then I can approve

**Review by glabute — CHANGES_REQUESTED** [review]
Still not quite there.  Problems with the folder names.  Top-level folder names in Assets need to match those in Samples source, so that I can copy them over when I change samples locally and want to commit.  Those folders should be:
- 2D Samples
- 3D Samples
- Shared Assets

**Comment by glabute** [comment]
Sample scripts are all broken.  Did the metafiles change?

image

**Comment by glabute** [comment]
metafiles are fine, it was a local corruption

**Review by glabute — APPROVED** [review]

**Pull request merged**